### ISSUE TITLE
Cannot handle generic associated type with `where` statement

### ISSUE THREAD

**Issue by ArcticLampyrid** [rust-lang/rustfmt#6274]
For code:
```rust
pub trait Foo {
    type Bar
    where
        Self: Sized;
}

impl Foo for u8 {
    // note: there is a space between `u16` and LineFeed
    type Bar = u16 
    where Self: Sized;
}
```

I got the error: 
```
$ cargo fmt
error[internal]: left behind trailing whitespace
 --> \\?\D:\foo\src\a.rs:9:9:19
  |
9 |     type Bar = u16 
  |                   ^
  |

warning: rustfmt has failed to format. See previous 1 errors.
```

**Comment by johnhuichen** [comment]
just wondering, is 

```rust
type Bar = u16 
    where Self: Sized;
```

really the right syntax?

**Comment by ArcticLampyrid** [comment]
@johnhuichen 
I did not check any RFC documents, but I believe it is one type of GATs.
It works on rustc 1.70, with no build error or runtime error.

**Comment by johnhuichen** [comment]
I am just thinking that `where Self:Sized` tells trait not to implement Type Bar if it's not sized. 

Since you already declared the type, does it still make sense to have `where Self:Size` in the implementation?

**Comment by ytmimi** [comment]
@ArcticLampyrid this one should be addressed once we merge https://github.com/rust-lang/rustfmt/pull/5887, which we're planning to do in the near future. 

@johnhuichen yes it's the correct syntax and it relates to https://github.com/rust-lang/rust/issues/89122. Also, see the [type alias docs of the rust reference](https://doc.rust-lang.org/reference/items/type-aliases.html#:~:text=Where%20clauses%20before%20the%20equals%20sign%20on%20a%20type%20alias%20in%20a%20trait%20impl%20(like%20type%20TypeAlias%3CT%3E%20where%20T%3A%20Foo%20%3D%20Bar%3CT%3E)%20are%20deprecated.%20Where%20clauses%20after%20the%20equals%20sign%20(like%20type%20TypeAlias%3CT%3E%20%3D%20Bar%3CT%3E%20where%20T%3A%20Foo)%20are%20preferred).

**Comment by johnhuichen** [comment]
> @ArcticLampyrid this one should be addressed once we merge #5887, which we're planning to do in the near future.
> 
> @johnhuichen yes it's the correct syntax and it relates to [rust-lang/rust#89122](https://github.com/rust-lang/rust/issues/89122). Also, see the [type alias docs of the rust reference](https://doc.rust-lang.org/reference/items/type-aliases.html#:~:text=Where%20clauses%20before%20the%20equals%20sign%20on%20a%20type%20alias%20in%20a%20trait%20impl%20(like%20type%20TypeAlias%3CT%3E%20where%20T%3A%20Foo%20%3D%20Bar%3CT%3E)%20are%20deprecated.%20Where%20clauses%20after%20the%20equals%20sign%20(like%20type%20TypeAlias%3CT%3E%20%3D%20Bar%3CT%3E%20where%20T%3A%20Foo)%20are%20preferred).

Thanks! I see what's going on now. The `where Self: Sized` in `impl` block is not necessarily needed but it's also allowed syntax

**Comment by ArcticLampyrid** [comment]
> Thanks! I see what's going on now. I think the `where Self: Sized` in `impl` block is allowed but not necessarily needed

It is not code in the real world, it's just a test case.


A more practical usage is:
```rust
impl<A> Foo<A> for MyType<A> {
    type Bar = u16 
    where
        A: Sized;
}
```
